### PR TITLE
fix auto reconnect after stop bug

### PR DESCRIPTION
**Pull request by realwanpengli**

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs [hunk 1]
```diff
@@ -72,7 +72,7 @@ private set
                 {
                     lock (_statusLock)
                     {
-                        if (_status != value)
+                        if (_status != value && _status != ServiceConnectionStatus.Terminated)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs]: why adding a status? a `_close` bool is enough?

Comment (realwanpengli) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs]: This is better? Terminated means the connection cannot reconnect anymore. It let others know the current connection's status

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs]: No. I don't think ServerConnection itself is aware of "reconnect" stuff. Introducing in a new concept makes it more complex. The container is where to determine how it handles the connections.

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs [hunk 2]
```diff
@@ -126,7 +133,10 @@ protected async Task StartCoreAsync(IServiceConnection connection, string target
             }
             finally
             {
-                await OnConnectionComplete(connection);
+                if (!_terminated)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs]: move this logic inside `OnConnectionComplete`

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/WeakServiceConnectionContainer.cs [hunk 3]
```diff
@@ -59,6 +59,8 @@ public override Task WriteAsync(ServiceMessage serviceMessage)
             return base.WriteAsync(serviceMessage);
         }
 
+        internal ServiceConnectionStatus GetServiceConnectionStatus() => GetStatus();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/WeakServiceConnectionContainer.cs]: Why? `Status` not enough?

File: test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs [hunk 4]
```diff
@@ -171,9 +171,9 @@ internal async Task SendToConnectionTest(ServiceTransportType serviceTransportTy
             try
             {
                 await RunTestCore(clientEndpoint, clientAccessTokens,
-                    async () => 
+                    async () =>
                     {
-                        var connectionId = await task.OrTimeout();
+                        var connectionId = await SignalR.Tests.Common.TaskExtensions.OrTimeout(task);
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs]: why change?

File: test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs [hunk 5]
```diff
@@ -217,6 +217,31 @@ internal async Task ConnectionJoinLeaveGroupTest(ServiceTransportType serviceTra
             }
         }
 
+        [ConditionalFact]
+        [SkipIfConnectionStringNotPresent]
+        internal async Task StopServiceHubContextTest()
+        {
+            using (StartVerifiableLog(out var loggerFactory, LogLevel.Debug))
+            {
+
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs]: remove extra lines

File: test/Microsoft.Azure.SignalR.Tests.Common/TaskExtensions.cs [hunk 6]
```diff
@@ -1,10 +1,13 @@
...
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System;
 using System.Diagnostics;
 using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
 
-namespace System.Threading.Tasks
+namespace Microsoft.Azure.SignalR.Tests.Common
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.Tests.Common/TaskExtensions.cs]: why change?

File: test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs (hunk 4 above)
Comment (realwanpengli) [reply test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs]: Management project also has a `TaskExtension`, conflict between both extensions.

File: src/Microsoft.Azure.SignalR.Management/ServiceHubContext.cs [hunk 7]
```diff
@@ -18,11 +19,15 @@ internal class ServiceHubContext : IServiceHubContext
 
         public IUserGroupManager UserGroups { get; }
 
+        // for test only
+        internal IServiceConnectionContainer ConnectionContainer { get; }
```
Comment (yanweizh) [inline src/Microsoft.Azure.SignalR.Management/ServiceHubContext.cs]: Internal above public

File: test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs [hunk 8]
```diff
@@ -217,6 +217,28 @@ internal async Task ConnectionJoinLeaveGroupTest(ServiceTransportType serviceTra
...
+                    {
+                        o.ConnectionString = TestConfiguration.Instance.ConnectionString;
+                        o.ConnectionCount = 1;
+                        o.ServiceTransportType = ServiceTransportType.Persistent;
+                    })
+                    .Build();
+                var serviceHubContext = await serviceManager.CreateHubContextAsync("hub", loggerFactory);
+                var connectionContainer = ((ServiceHubContext)serviceHubContext).ConnectionContainer;
+                await serviceHubContext.DisposeAsync();
+                await Task.Delay(500);
```
[the remark this answers is not in the document]
Comment (realwanpengli) [reply test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs]: make sure it won't reconnect later

[the remark this answers is not in the document]
Comment (vicancy) [reply test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs]: it makes tests 0.5 seconds longer 😅

File: src/Microsoft.Azure.SignalR.Management/ServiceHubContext.cs [hunk 9]
```diff
@@ -3,14 +3,19 @@
...
     internal class ServiceHubContext : IServiceHubContext
     {
-        private IHubContext<Hub> _hubContext;
-        private ServiceProvider _serviceProvider;
+        private readonly IHubContext<Hub> _hubContext;
+        
+        private readonly ServiceProvider _serviceProvider;
+
+        // for test only
+        internal IServiceConnectionContainer ConnectionContainer { get; }
```
[the remark this answers is not in the document]
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Management/ServiceHubContext.cs]: init in .ctor is better as it is only set once

File: test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (realwanpengli) [reply test/Microsoft.Azure.SignalR.E2ETests/Management/ServiceHubContextE2EFacts.cs]: looks in other tests, we only use `TimeSpan.Fromxxx` if the unit is not milliseceond

File: src/Microsoft.Azure.SignalR.Management/ServiceHubContext.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (realwanpengli) [reply src/Microsoft.Azure.SignalR.Management/ServiceHubContext.cs]: Can we create a inner class and make all metheds and fields only placed in this inner class? this makes the outer class clean.

**Pull request merged**